### DISCUSSION TITLE
Migrate from `__dataclass_transform__` to `typing.dataclass_transform`

### DISCUSSION THREAD

**Discussion by multimeric** [pydantic/pydantic#4238] [Feature Request]
As of June 2022, [PEP 681 has been accepted](https://mail.python.org/archives/list/python-dev@python.org/message/R4A2IYLGFHKFDYJPSDA5NFJ6N7KRPJ6D/). Therefore, it would be good if Pydantic could migrate from `__dataclass_transform__` (which was only ever a [temporary measure](https://github.com/microsoft/pyright/blob/main/specs/dataclass_transforms.md#alternate-form)) to `typing.dataclass_transform`, which would satisfy the PEP specification.

This would allow type checkers that support PEP 681 but not the intermediate measure to appropriately type check and autocomplete pydantic models.

**Comment by samuelcolvin** [comment]
But which versions of python is it available in? Is it available in typing-extensions? If so, is it recognized by all applicable type checkers?

What do well do about mypy where we have the dedicated extension?

**Reply by hmc-cs-mdrissi** [reply]
3.11 for python version. It is available in typing extensions. It is supported by pyright/pyre, while it is a no-op for the other type checkers. It won’t help mypy but also won’t hurt it. Until mypy has support for it the pydantic extension should continue to be used.
